Adding TypeScript to "Getting Started" (#1800)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/js/getting-started/browser.md

aliases: [/docs/js/getting_started/browser]
---

This guide uses the example application in HTML & typescript/javascript provided below, but the steps to instrument your own application should be broadly the same.

> ## Example Application

Reviewer: Before this section we should have a section entitled prerequisites

> ## Prerequisites

You'll probably want to mention Node.js requirements.

In the edits presented in this file, there seems to be an assumption that `tsc` is installed and available. If that is the case, then it should be mentioned. Otherwise, (or maybe regardless), readers to be referred to https://www.typescriptlang.org/download if they need to install TypeScript.

I particularly like the NPM based install, but then that would require using `npx tsc` instead of `tsc` below. WDYT @svrnm?

Reviewer 2: Is `npx` installed by default? I am not 100% sure anymore, but if so, yes, let's reduce the steps people have to take.

Author: Yes the `npx` gets installed by default with npm versions above 5 I guess

> While this guide uses the example application presented below, the steps to instrument your own application should be similar. Please make sure you have required [Typescript](https://www.typescriptlang.org/download) or [NodeJs](https://nodejs.org/en/download/) installed on your system.

Reviewer: ```suggestion
While this guide uses the example application presented below, the steps to
instrument your own application should be similar.

## Prerequisites

Ensure that you have the following installed locally:

- [Node.js](https://nodejs.org/en/download/)
- [TypeScript](https://www.typescriptlang.org/download), if you will be using TypeScript.
```

### Installation

Let's initialize the the package and config files for the example project.

{{< tabs TypeScript JavaScript >}}

{{< tab >}}
$ tsc --init
{{< /tab >}}

{{< tab >}}
$ npm init -y
{{< /tab >}}

{{< /tabs >}}

> 

Reviewer: As I mentioned before, it doesn't make sense to use tabs here. For context, see the following thread: https://github.com/open-telemetry/opentelemetry.io/pull/1800#discussion_r989541203

To create traces in the browser, you will need `@opentelemetry/sdk-trace-web`, and the instrumentation `@opentelemetry/instrumentation-document-load`:

```shell

> $ npm install  @opentelemetry/api @opentelemetry/sdk-trace-web @opentelemetry/instrumentation-document-load @opentelemetry/context-zone

Reviewer: - If @svrnm et all feel that you want to keep the init command then add this back.
- Also, drop the `$` since this is a `shell` block

```suggestion
npm init -y
npm install  @opentelemetry/api @opentelemetry/sdk-trace-web @opentelemetry/instrumentation-document-load @opentelemetry/context-zone
```

Reviewer 2: as said, I prefer having that npm init -y there, it's not really necessary, but it helps with a few corner cases, especially when you have node_modules anywhere in a parent directory.

@cartermp , wdyt?

Reviewer: I'm ok with the npm init.

> ```bash

Reviewer 2: ```suggestion
```shell
```

```

### Initialization and Configuration

> 

Reviewer: ```suggestion

If you are coding in TypeScript, then run the following command: `tsc --init`.

```

Of course, you can also put the command in a `shell` code-block .

Create an empty code file named `document-load` with a `.ts` or `.js` extension , as appropriate, based on the language you've chosen to write your app in. Add the following code to your HTML right before the `</body>` closing tag:

{{< tabs TypeScript JavaScript >}}

{{< tab lang="html">}}
<script type="module" src="document-load.ts"></script>
{{< /tab >}}

{{< tab lang="html">}}
<script type="module" src="document-load.js"></script>
{{< /tab >}}

{{< /tabs >}}

We will add some code that will trace the document load timings and output those as OpenTelemetry Spans.

### Creating a Tracer Provider

Add the following code to the `document-load.ts|js` to create a tracer provider, which brings the instrumentation to trace document load:

```js

> /* document-load.ts|js file (Code Snippet is similar for both the languages)*/

Reviewer: ```suggestion
/* document-load.ts|js file - the code is the same for both the languages */
```

To create traces in the browser, you will need `@opentelemetry/sdk-trace-web`, and the instrumentation `@opentelemetry/instrumentation-document-load`:

{{< tabs TypeScript JavaScript >}}

{{< tab >}}
tsc --init
npm install --save @opentelemetry/api @opentelemetry/sdk-trace-web @opentelemetry/instrumentation-document-load @opentelemetry/context-zone
{{< /tab >}}

{{< tab >}}
npm init -y
npm install --save @opentelemetry/api @opentelemetry/sdk-trace-web @opentelemetry/instrumentation-document-load @opentelemetry/context-zone
{{< /tab >}}

> {{< /tabs >}}

Reviewer 2: > Drop the npm init -y, AFAIK, you don't need it for this simple example app.

I introduced that a while back into the nodejs getting started because I had a node_modules in a parent directory and node is then putting everything there instead of the local directory. Although this is expected behaviour by npm, it's confusing, having a basic `package.json` fixes that. The alternative is `echo {} > package.json`.

> {{< /tabs >}}

Author: I will just keep the `npm init -y` and `tsc --init` commands for now

> {{< /tabs >}}

Reviewer: > I introduced that a while back into the nodejs getting started because I had a node_modules in a parent directory and node is then putting everything there instead of the local directory. Although this is expected behaviour by npm, it's confusing, having a basic `package.json` fixes that.

How likely is it that a first-time user will have a node_modules folder in the parent directory? I'd prefer that we drop the `npm init -y` command, but I'm not strongly against it. 

@kaiwalyakoparkar - all of the other points I made are still relevant. I'll point this out again in my next review.

### Initialization and Configuration

> Create a empty file called `document-load.ts` or `document-load.js` (Based on if you are using typescript or javascript) and add the following code to your html right before the body end tag:

Reviewer 2: And from there on we call it "code file"? That would work for me 👍

{{< tabs TypeScript JavaScript >}}

{{< tab >}}
<script type="module" src="document-load.ts"></script>
{{< /tab >}}

{{< tab >}}
<script type="module" src="document-load.js"></script>
{{< /tab >}}

{{< /tabs >}}

We will add some code that will trace the document load timings and output those as OpenTelemetry Spans.

### Creating a Tracer Provider

> Add the following code to the `document-load.ts` or `document-load.js` to create a tracer provider, which brings the instrumentation to trace document load:

Reviewer: I agree. How about:

```suggestion
Add the following code to the `document-load.*` you created in the previous step. The code creates a tracer provider, which brings the instrumentation to trace document load:
```

> Add the following code to the `document-load.ts` or `document-load.js` to create a tracer provider, which brings the instrumentation to trace document load:

Reviewer 2: I like that 👍 or do people understand `document-load.ts|js`?

> Add the following code to the `document-load.ts` or `document-load.js` to create a tracer provider, which brings the instrumentation to trace document load:

Author: Okay I will try implementing `document-load.ts|js`.

{{< tabs TypeScript JavaScript >}}

{{< tab >}}
/* document-load.ts */
import { WebTracerProvider } from '@opentelemetry/sdk-trace-web';
import { DocumentLoadInstrumentation } from '@opentelemetry/instrumentation-document-load';
import { ZoneContextManager } from '@opentelemetry/context-zone';
import { registerInstrumentations } from '@opentelemetry/instrumentation';

> const provider:WebTracerProvider = new WebTracerProvider();

Reviewer: Is this type declaration really necessary? Can't TS infer the type automatically? (I haven't worked with TS in a while so I don't know if it can do something like that now -- also, I might be confusing it with [Dart](https://dart.dev/).)

If the type declaration isn't necessary, then the two code blocks are identical, and so we should avoid tabs, just present the code that is valid for both. My 2 cents.

Author: @chalin you are right I will remove the tabbing and make it as a single code then

```shell
npm init -y
npm install  @opentelemetry/api @opentelemetry/sdk-trace-web @opentelemetry/instrumentation-document-load @opentelemetry/context-zone
```

### Initialization and Configuration

If you are coding in TypeScript, then run the following command: 
```shell
tsc --init
```

> Create an empty code file named `document-load` with a `.ts` or `.js` extension , as appropriate, based on the language you've chosen to write your app in. Add the following code to your HTML right before the `</body>` closing tag:

Reviewer 2: ```suggestion
Create an empty code file named `document-load` with a `.ts` or `.js` extension, as appropriate, based on the language you've chosen to write your app in. Add the following code to your HTML right before the `</body>` closing tag:
```

Run

```shell

> $ npx parcel index.html

Reviewer: Don't put a `$` for `shell` environments (which is the proper code-block lang to use when there is no command output):

```suggestion
npx parcel index.html
```

You may also want to use the `BatchSpanProcessor` to export spans in batches in order to more efficiently use resources.

To export traces to the console, modify `document-load.ts|js` so that it matches the following code snippet:

```js

> /* document-load.ts|js file (Code Snippet is similar for both the languages) */

Reviewer: ```suggestion
/* document-load.ts|js file - the code is the same for both the languages */
```

You may also want to use the `BatchSpanProcessor` to export spans in batches in order to more efficiently use resources.

> To export traces to the console, modify `document-load.ts` or `document-load.js` so that it matches the following code snippet:

Reviewer: ```suggestion
To export traces to the console, modify `document-load.*` so that it matches the following code snippet:
```

{{< tabs TypeScript JavaScript >}}

{{< tab >}}
/* document-load.ts */
import { ConsoleSpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { WebTracerProvider } from '@opentelemetry/sdk-trace-web';
import { DocumentLoadInstrumentation } from '@opentelemetry/instrumentation-document-load';
import { ZoneContextManager } from '@opentelemetry/context-zone';
import { registerInstrumentations } from '@opentelemetry/instrumentation';

> const provider:WebTracerProvider = new WebTracerProvider();

Reviewer: Same question as above. Is this type specifier necessary?
